core/src/main/scala/org/apache/spark/util/io/ChunkedByteBufferOutputStream.scala

```diff
@@ -18,19 +18,25 @@
 package org.apache.spark.util.io
 
 import java.io.OutputStream
+import java.nio.ByteBuffer
 
 import scala.collection.mutable.ArrayBuffer
 
+import org.apache.spark.storage.StorageUtils
 
 /**
  * An OutputStream that writes to fixed-size chunks of byte arrays.
  *
  * @param chunkSize size of each chunk, in bytes.
  */
-private[spark]
-class ByteArrayChunkOutputStream(chunkSize: Int) extends OutputStream {
+private[spark] class ChunkedByteBufferOutputStream(
+    chunkSize: Int,
+    allocator: Int => ByteBuffer)
```

Comment:
seems like this will always be either `ByteBuffer.allocate` or `Platform.allocateDirectBuffer`. It might be better to pass in a boolean here instead `offHeap: Boolean = false` or something and do
```
def allocate(size: Int): ByteBuffer = {
  if (offHeap) Platform.allocateDirectBuffer(size) else ByteBuffer.allocate(size)
}
```
then you don't need to pass a closure around